app/routines/export_users_last_login_date.rb

```diff
@@ -0,0 +1,52 @@
+class ExportUsersLastLoginDate
+  lev_routine
+
+  protected
+
+  def exec(delete_after: true)
+    outputs[:filename] = filename
+    outputs[:info] = info
+    generate_csv
+    remove_exported_files if delete_after
+  end
+
+  def info
+    output_users = []
+    User.preload(:contact_infos).find_each do |user|
+      output_users << Hashie::Mash.new({
+        emails: user.contact_infos.map(&:value).join(", "),
+        last_login_at: SecurityLog.sign_in_successful
+                                    .where(user_id: user.id).try(:last).try(:created_at).try(:strftime, "%m/%d/%Y %I:%M%p %Z")
```

Comment:
I suggest using `&.`: `&.last&.created_at&.strftime(...)` etc.
Also I think you want `first` instead of `last` because the `default_scope` for SecurityLog is `created_at DESC`.

Reply:
**Great** catch with the ordering!! I didn't think too hard about that.

Reply:
I totally did not know about that Safe Navigation Operator_. 💯 Fixed.